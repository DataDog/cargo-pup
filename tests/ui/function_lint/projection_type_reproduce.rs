//@compile-flags: --crate-name projection_type_reproduce
//@compile-flags: --crate-type lib

use std::fmt;

// A trait with an associated error type
trait MyTrait {
    type Error: std::error::Error + Default;
}

// A concrete error type
#[derive(Debug, Default)]
struct MyError;

impl fmt::Display for MyError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "MyError")
    }
}

impl std::error::Error for MyError {}

impl MyTrait for () {
    type Error = MyError;
}

// This is the function under lint
fn test<T: MyTrait>() -> Result<(), T::Error> //~ ERROR: Function exceeds maximum length of 1 lines with 4 line
where
    T::Error: std::error::Error + Default,
{
    // This triggers the check: does `<T as MyTrait>::Error` implement `Error`?
    Err(T::Error::default())
}