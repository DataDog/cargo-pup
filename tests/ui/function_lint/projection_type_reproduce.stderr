error: Function exceeds maximum length of 1 lines with 4 lines
  --> tests/ui/function_lint/projection_type_reproduce.rs:28:1
   |
LL | fn test<T: MyTrait>() -> Result<(), T::Error>
   | ^^^^^^^^^
   |
   = help: Consider breaking this function into smaller parts
   = note: Applied by cargo-pup rule 'projection_type_error_impl_test'.
   = note: `#[deny(function_lint)]` on by default

error: aborting due to 1 previous error

